### ISSUE TITLE
Type narrowing based on iterator `done` value?

### ISSUE THREAD

**Issue by jeremymeng** [microsoft/TypeScript#31952]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:**

**Code**

```ts
interface MyIterator<T> {
    next():
      | {
          done: false;
          value: T;
        }
      | {
          done: true;
          value: undefined;
        };
    [Symbol.iterator](): MyIterator<T>;
}

interface Item { name: string }

function m(iter: MyIterator<Item>) {
    for (const it of iter) {
        console.log(it.name);           // Error here complaining about 'it': Object is possibly 'undefined'.ts(2532)
    }
}
```

**Expected behavior:**

**Actual behavior:**

**Playground Link:** 
https://typescript-play.js.org/?downlevelIteration=true#code/JYOwLgpgTgZghgYwgAgLIE8CSkpzAeygB4AVAPmQG8AoZO5ECADzAAoBKALlvroB8qPXrwAm+Rp2TwANgGcIAbiHC6ANzjSArhEkklKugF9l-QQfpiJyMFG37zydVp3JNIERBigII+yuO8fsgA2gDK6AC2AEb40gB0wDh4hAC6HJIY2NDJxORKxtSgOPBIyFkRVAxwES6yNqAA5sgFMG4IYMDiyBGsidAZWEkExOVk7Gb0MITIrAjidciJyPgwizjjNOZzILKxEHHS+A29YHEg1RDsCg50APS3yACiUFDTABbQKHMRAA7ScKBGsg4DFNGBkAByRIQyQAeSiACsIO1FrJkD98LJZMAotJ0JC3B4vIwRBC4mBZKwAEwAVgAzFT2NRlMZDEA

**Related Issues:** 
https://github.com/microsoft/TypeScript/issues/11450 seems related. but here even with `--strictNullChecks` same error is reported here.
